File: scripts/_template_helpers/sync.py
#!/usr/bin/env python3
"""Phase 3 Commit A sync operations.

Subcommands:
  apply-overwrite     --new-tree <dir> --user-tree <dir> --rel <relpath>
  apply-three-way     --old-tree <dir> --new-tree <dir> --user-tree <dir> --rel <relpath>
  apply-attributes    --old-tree <dir> --new-tree <dir> --user-tree <dir> --rel <relpath>
                      [--accept-attribute-changes]
  apply-new-file      --new-tree <dir> --user-tree <dir> --rel <relpath>
                      --decision <overwrite|skip|mark-as-user-deleted>
  apply-deletion      --user-tree <dir> --rel <relpath> --decision <remove|preserve-local>
  has-conflict-markers --tree <dir> --paths <relpath>...
"""
from __future__ import annotations

import argparse
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path


def cmd_apply_overwrite(args: argparse.Namespace) -> int:
    src = args.new_tree / args.rel
    dst = args.user_tree / args.rel
    if not src.is_file():
        print(f"source not found: {src}", file=sys.stderr)
        return 1
    dst.parent.mkdir(parents=True, exist_ok=True)
    shutil.copyfile(src, dst)
    # Preserve mode bits (executable).
    dst.chmod(src.stat().st_mode)
    return 0


def cmd_apply_three_way(args: argparse.Namespace) -> int:
    base = args.old_tree / args.rel
    new = args.new_tree / args.rel
    cur = args.user_tree / args.rel
    if not new.is_file():
        return 0  # not in new tree — nothing to do
    if not cur.is_file():
        # User deleted the file; treat as new-file-with-base.
        cur.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(new, cur)
        return 0
    if not base.is_file():
        # No ancestor — fall back to overwrite-with-warning.
        shutil.copyfile(new, cur)
        print(f"warn: no ancestor for {args.rel}; overwrote", file=sys.stderr)
        return 0
    rc = subprocess.call(
        ["git", "merge-file", "--diff3",
         "-L", "current", "-L", "base", "-L", "new",
         str(cur), str(base), str(new)]
    )
    return rc


def cmd_apply_attributes(args: argparse.Namespace) -> int:
    rel = args.rel
    base = args.old_tree / rel
    new = args.new_tree / rel
    cur = args.user_tree / rel
    if not new.is_file():
        return 0
    # 3-way merge first (or fall-back overwrite if no base / no cur).
    if not cur.is_file():
        cur.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(new, cur)
    elif base.is_file():
        rc = subprocess.call(
            ["git", "merge-file", "--diff3",
             "-L", "current", "-L", "base", "-L", "new",
             str(cur), str(base), str(new)]
        )
        if rc != 0:
            # Conflict; let orchestrator's marker check halt.
            return rc
    else:
        shutil.copyfile(new, cur)

    # Attr audit: compare OLD .gitattributes vs merged .gitattributes for each tracked file.
    # Run check-attr from a clean tmp git repo so workdir .gitattributes doesn't pollute.
    tracked = subprocess.run(
        ["git", "-C", str(args.user_tree), "ls-files"],
        capture_output=True, text=True
    ).stdout.splitlines()

    base_attrs = str(base) if base.is_file() else os.devnull
    new_attrs = str(cur)

    flipped = False
    with tempfile.TemporaryDirectory() as td:
        subprocess.run(["git", "-C", td, "init", "-q"], check=True)
        env = {**os.environ, "GIT_ATTR_NOSYSTEM": "1"}
        for p in tracked:
            old_out = subprocess.run(
                ["git", "-C", td, "-c", f"core.attributesfile={base_attrs}",
                 "check-attr", "-a", "--", p],
                capture_output=True, text=True, env=env,
            ).stdout
            new_out = subprocess.run(
                ["git", "-C", td, "-c", f"core.attributesfile={new_attrs}",
                 "check-attr", "-a", "--", p],
                capture_output=True, text=True, env=env,
            ).stdout
            if old_out != new_out:
                old_enc = "filter: git-crypt" in old_out
                new_enc = "filter: git-crypt" in new_out
                if old_enc != new_enc:
                    flipped = True
                    print(f"PLAN|attribute-change|{p}|filter-flipped", file=sys.stderr)
    if flipped and not args.accept_attribute_changes:
        print("Encryption pattern change detected. Re-run with --accept-attribute-changes.",
              file=sys.stderr)
        return 1
    return 0


def cmd_apply_new_file(args: argparse.Namespace) -> int:
    src = args.new_tree / args.rel
    dst = args.user_tree / args.rel
    if args.decision == "skip":
        return 0
    if args.decision == "overwrite":
        if not src.is_file():
            print(f"source not found: {src}", file=sys.stderr)
            return 1
        dst.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(src, dst)
        dst.chmod(src.stat().st_mode)
        return 0
    if args.decision == "mark-as-user-deleted":
        # Caller (orchestrator) records the deletion in template.json.deleted[]
        # via state.py add-deleted-pending. Nothing to do on disk.
        return 0
    print(f"unknown decision: {args.decision}", file=sys.stderr)
    return 2


def cmd_apply_deletion(args: argparse.Namespace) -> int:
    target = args.user_tree / args.rel
    if args.decision == "preserve-local":
        return 0
    if args.decision == "remove":
        if target.is_file():
            target.unlink()
        return 0
    print(f"unknown decision: {args.decision}", file=sys.stderr)
    return 2


def cmd_has_conflict_markers(args: argparse.Namespace) -> int:
    found = False
    for p in args.paths:
        full = args.tree / p
        if not full.is_file():
            continue
        with full.open("rb") as f:
            for line in f:
                if line.startswith(b"<<<<<<<"):
                    found = True
                    print(p)
                    break
    return 1 if found else 0


def _add_overwrite(sub):
    p = sub.add_parser("apply-overwrite")
    p.add_argument("--new-tree", required=True, type=Path)
    p.add_argument("--user-tree", required=True, type=Path)
    p.add_argument("--rel", required=True)


def _add_three_way(sub):
    p = sub.add_parser("apply-three-way")
    p.add_argument("--old-tree", required=True, type=Path)
    p.add_argument("--new-tree", required=True, type=Path)
    p.add_argument("--user-tree", required=True, type=Path)
    p.add_argument("--rel", required=True)


def _add_attributes(sub):
    p = sub.add_parser("apply-attributes")
    p.add_argument("--old-tree", required=True, type=Path)
    p.add_argument("--new-tree", required=True, type=Path)
    p.add_argument("--user-tree", required=True, type=Path)
    p.add_argument("--rel", required=True)
    p.add_argument("--accept-attribute-changes", action="store_true")


def _add_new_file(sub):
    p = sub.add_parser("apply-new-file")
    p.add_argument("--new-tree", required=True, type=Path)
    p.add_argument("--user-tree", required=True, type=Path)
    p.add_argument("--rel", required=True)
    p.add_argument("--decision", required=True,
                   choices=["overwrite", "skip", "mark-as-user-deleted"])


def _add_deletion(sub):
    p = sub.add_parser("apply-deletion")
    p.add_argument("--user-tree", required=True, type=Path)
    p.add_argument("--rel", required=True)
    p.add_argument("--decision", required=True, choices=["remove", "preserve-local"])


def _add_has_markers(sub):
    p = sub.add_parser("has-conflict-markers")
    p.add_argument("--tree", required=True, type=Path)
    p.add_argument("--paths", required=True, nargs="+")


def main() -> int:
    parser = argparse.ArgumentParser()
    sub = parser.add_subparsers(dest="cmd", required=True)

    _add_overwrite(sub)
    _add_three_way(sub)
    _add_attributes(sub)
    _add_new_file(sub)
    _add_deletion(sub)
    _add_has_markers(sub)

    args = parser.parse_args()
    dispatch = {
        "apply-overwrite": cmd_apply_overwrite,
        "apply-three-way": cmd_apply_three_way,
        "apply-attributes": cmd_apply_attributes,
        "apply-new-file": cmd_apply_new_file,
        "apply-deletion": cmd_apply_deletion,
        "has-conflict-markers": cmd_has_conflict_markers,
    }
    return dispatch[args.cmd](args)


if __name__ == "__main__":
    sys.exit(main())

File: internal/template/sync.go
package template

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

// SyncGit captures the git operations Phase 3 Commit A needs. The bash +
// Python oracle drives `git merge-file --diff3`, `git ls-files`, `git
// init`, and `git check-attr`. Each method returns (code, err) for
// commands run for side-effects, or (out, code, err) for stdout-bearing
// commands. A non-nil err signals the binary is missing entirely.
type SyncGit interface {
	// MergeFile shells `git merge-file --diff3 -L current -L base -L
	// new <cur> <base> <new>`. The cur path is mutated in place. Exit
	// 0 = clean merge; >0 = number of conflicts.
	MergeFile(cur, base, newFile string) (code int, err error)
	// LsFilesIn shells `git -C <dir> ls-files`. Returns the raw newline-
	// separated stdout.
	LsFilesIn(dir string) (out string, code int, err error)
	// CheckAttrAll runs `git check-attr -a -- <path>` against an
	// isolated repo configured with core.attributesfile=<attrFile> and
	// GIT_ATTR_NOSYSTEM=1. The Python oracle creates a fresh tempdir
	// for this; production adapters should mirror that. attrFile may
	// be the literal "/dev/null" sentinel to indicate "no .gitattributes".
	CheckAttrAll(attrFile, path string) (out string, err error)
}

// SyncDecision is the user-supplied decision for new_file +
// deletion-in-template plan rows.
type SyncDecision string

const (
	SyncDecisionOverwrite          SyncDecision = "overwrite"
	SyncDecisionSkip               SyncDecision = "skip"
	SyncDecisionMarkAsUserDeleted  SyncDecision = "mark-as-user-deleted"
	SyncDecisionRemove             SyncDecision = "remove"
	SyncDecisionPreserveLocal      SyncDecision = "preserve-local"
)

// ApplyOverwrite copies newTree/rel to userTree/rel, preserving the
// source file mode. Mirrors `sync.py apply-overwrite`.
func ApplyOverwrite(newTree, userTree, rel string) error {
	src := filepath.Join(newTree, rel)
	dst := filepath.Join(userTree, rel)
	si, err := os.Stat(src)
	if err != nil || si.IsDir() {
		return fmt.Errorf("source not found: %s", src)
	}
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	if err := copyFile(src, dst); err != nil {
		return err
	}
	return os.Chmod(dst, si.Mode().Perm())
}

// ApplyThreeWay performs the three-way merge for rel. Returns the
// merge-file exit code (0 = clean, >0 = number of conflicts) so callers
// can record / surface the marker check failure mode. Mirrors
// `sync.py apply-three-way`.
func ApplyThreeWay(g SyncGit, oldTree, newTree, userTree, rel string) (int, error) {
	base := filepath.Join(oldTree, rel)
	newFile := filepath.Join(newTree, rel)
	cur := filepath.Join(userTree, rel)
	if !isFile(newFile) {
		return 0, nil // not in new tree — nothing to do
	}
	if !isFile(cur) {
		// User deleted the file; treat as new-file-with-base.
		if err := os.MkdirAll(filepath.Dir(cur), 0o755); err != nil {
			return 0, err
		}
		return 0, copyFile(newFile, cur)
	}
	if !isFile(base) {
		// No ancestor — fall back to overwrite-with-warning. Caller
		// surfaces the warning to stderr.
		return 0, copyFile(newFile, cur)
	}
	if g == nil {
		return 0, errors.New("sync: git adapter is nil")
	}
	return g.MergeFile(cur, base, newFile)
}

// ApplyAttributesResult carries the cross-cutting state Phase-3-Commit-A
// inspects after `apply-attributes` completes.
type ApplyAttributesResult struct {
	// MergeCode is the underlying merge-file exit code, or 0 if no
	// merge ran (e.g. user deleted the file).
	MergeCode int
	// Flipped is true iff the OLD vs NEW .gitattributes encryption
	// filter changed for at least one tracked path.
	Flipped bool
	// FlippedPaths is the list of paths whose encryption filter
	// flipped (used to emit PLAN|attribute-change|... lines).
	FlippedPaths []string
}

// ApplyAttributes mirrors `sync.py apply-attributes`. It runs the
// three-way merge, then audits the OLD vs merged .gitattributes for
// flipped git-crypt filter assignments.
//
// When acceptAttrChanges is false and any flip is detected, the caller
// should treat that as the halt condition (matches Python `return 1`).
func ApplyAttributes(g SyncGit, oldTree, newTree, userTree, rel string, acceptAttrChanges bool) (*ApplyAttributesResult, error) {
	res := &ApplyAttributesResult{}
	base := filepath.Join(oldTree, rel)
	newFile := filepath.Join(newTree, rel)
	cur := filepath.Join(userTree, rel)
	if !isFile(newFile) {
		return res, nil
	}
	if !isFile(cur) {
		if err := os.MkdirAll(filepath.Dir(cur), 0o755); err != nil {
			return res, err
		}
		if err := copyFile(newFile, cur); err != nil {
			return res, err
		}
	} else if isFile(base) {
		if g == nil {
			return res, errors.New("sync: git adapter is nil")
		}
		code, err := g.MergeFile(cur, base, newFile)
		if err != nil {
			return res, err
		}
		res.MergeCode = code
		if code != 0 {
			// Conflict; let orchestrator's marker check halt.
			return res, nil
		}
	} else {
		if err := copyFile(newFile, cur); err != nil {
			return res, err
		}
	}

	if g == nil {
		return res, errors.New("sync: git adapter is nil")
	}

	// Tracked files audit.
	out, _, err := g.LsFilesIn(userTree)
	if err != nil {
		return res, err
	}
	tracked := splitLines(out)

	baseAttrs := base
	if !isFile(base) {
		baseAttrs = os.DevNull
	}
	newAttrs := cur

	for _, p := range tracked {
		if p == "" {
			continue
		}
		oldOut, err := g.CheckAttrAll(baseAttrs, p)
		if err != nil {
			return res, err
		}
		newOut, err := g.CheckAttrAll(newAttrs, p)
		if err != nil {
			return res, err
		}
		if oldOut != newOut {
			oldEnc := strings.Contains(oldOut, "filter: git-crypt")
			newEnc := strings.Contains(newOut, "filter: git-crypt")
			if oldEnc != newEnc {
				res.Flipped = true
				res.FlippedPaths = append(res.FlippedPaths, p)
			}
		}
	}
	if res.Flipped && !acceptAttrChanges {
		return res, ErrAttributeChangeRejected
	}
	return res, nil
}

// ErrAttributeChangeRejected is the sentinel ApplyAttributes returns
// when an encryption-filter flip is detected without
// --accept-attribute-changes. Callers should print the same hint Python
// does ("Re-run with --accept-attribute-changes.") and exit 1.
var ErrAttributeChangeRejected = errors.New("encryption pattern change detected; re-run with --accept-attribute-changes")

// ApplyNewFile mirrors `sync.py apply-new-file`. Decision must be one
// of SyncDecisionOverwrite / Skip / MarkAsUserDeleted.
func ApplyNewFile(newTree, userTree, rel string, decision SyncDecision) error {
	switch decision {
	case SyncDecisionSkip:
		return nil
	case SyncDecisionOverwrite:
		src := filepath.Join(newTree, rel)
		dst := filepath.Join(userTree, rel)
		si, err := os.Stat(src)
		if err != nil || si.IsDir() {
			return fmt.Errorf("source not found: %s", src)
		}
		if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
			return err
		}
		if err := copyFile(src, dst); err != nil {
			return err
		}
		return os.Chmod(dst, si.Mode().Perm())
	case SyncDecisionMarkAsUserDeleted:
		// On-disk no-op. Caller records the deletion in template.json.
		return nil
	default:
		return fmt.Errorf("unknown decision: %s", decision)
	}
}

// ApplyDeletion mirrors `sync.py apply-deletion`. Decision must be
// SyncDecisionRemove or SyncDecisionPreserveLocal.
func ApplyDeletion(userTree, rel string, decision SyncDecision) error {
	switch decision {
	case SyncDecisionPreserveLocal:
		return nil
	case SyncDecisionRemove:
		target := filepath.Join(userTree, rel)
		if isFile(target) {
			return os.Remove(target)
		}
		return nil
	default:
		return fmt.Errorf("unknown decision: %s", decision)
	}
}

// HasConflictMarkers mirrors `sync.py has-conflict-markers`. It scans
// each tree/path file for the byte sequence "<<<<<<<" at line start.
// Returns the list of paths with markers; an empty slice means clean.
func HasConflictMarkers(tree string, paths []string) ([]string, error) {
	var hits []string
	for _, p := range paths {
		full := filepath.Join(tree, p)
		fi, err := os.Stat(full)
		if err != nil {
			if errors.Is(err, os.ErrNotExist) {
				continue
			}
			return nil, err
		}
		if fi.IsDir() {
			continue
		}
		f, err := os.Open(full)
		if err != nil {
			return nil, err
		}
		// Match Python's iter-over-bytes-line approach.
		scanner := bufio.NewScanner(f)
		buf := make([]byte, 0, 64*1024)
		scanner.Buffer(buf, 1024*1024)
		for scanner.Scan() {
			line := scanner.Bytes()
			if strings.HasPrefix(string(line), "<<<<<<<") {
				hits = append(hits, p)
				break
			}
		}
		_ = f.Close()
	}
	return hits, nil
}

// copyFile is a streaming file copy that does NOT preserve mode bits;
// callers chmod the destination explicitly. Mirrors shutil.copyfile.
func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer func() { _ = in.Close() }()
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	if _, err := io.Copy(out, in); err != nil {
		_ = out.Close()
		return err
	}
	return out.Close()
}

// splitLines splits s on '\n', dropping a trailing empty entry produced
// by an "...\n"-terminated string.
func splitLines(s string) []string {
	if s == "" {
		return nil
	}
	parts := strings.Split(s, "\n")
	if len(parts) > 0 && parts[len(parts)-1] == "" {
		parts = parts[:len(parts)-1]
	}
	return parts
}
